README.md
# query-validation

[![Build Status](https://travis-ci.com/stanford-oval/query-validation.svg?branch=master)](https://travis-ci.com/stanford-oval/query-validation) [![Coverage Status](https://coveralls.io/repos/github/stanford-oval/query-validation/badge.svg?branch=master)](https://coveralls.io/github/stanford-oval/query-validation?branch=master) [![Dependency Status](https://david-dm.org/stanford-oval/query-validation/status.svg)](https://david-dm.org/stanford-oval/query-validation) [![Language grade: JavaScript](https://img.shields.io/lgtm/grade/javascript/g/stanford-oval/query-validation.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/stanford-oval/query-validation/context:javascript)

A small library to do server-side validation of GET and POST parameters. It is designed
to work with express, qs and body-parser.

## Usage

```javascript
const qv = require('query-validation');
const express = require('express');

const app = express();
app.get('/foo', qv.validateGET({
   param1: 'string',
   param2: 'number'
}), (req, res, next) => {
   // use req.query.param1 and req.query.param2 safely
});
```

## API

### validateGET(keys, options)

Returns a middleware function that validates the query in the request (`req.query`) according to the passed keys. The following types are accepted:

- `string`: must be a non-empty string, with no ASCII control characters except ordinary whitespace
- `number`: must be a finite, not-NaN number (either a finite JS number, or a string representing a finite number in decimal notation)
- `integer`: must be an integer number (either a JS number that is an integer, or a string representing an integer in decimal notation)
- `boolean`: must be a JS boolean, the string '1', the empty string, or undefined (missing) (this is designed to handle HTML checkboxes)
- `object`: must be a plain object (not a primitive or Array)
- `array`: must be an Array
- `null`: must be the JS `null` (this is only meaningful for JSON input)
- a RegExp object: must be a string conforming to the passed-in regular expression

All types may be prefixed with `?` to indicate that the value is optional, in which case `undefined` is accepted.

If the query is not valid, the request will fail with a `ValidationError`, passed to the `next` handler of the middleware. The error can
be handled using express standard error handling middleware, or it will cause an Internal Server Error.

### validatePOST(keys, options)

Similar to `validateGET`, but validates `req.body` instead. This is designed to work with `bodyParser.urlencoded`, `bodyParser.json` or similar
middlewares.

Options:
- `options.accept`: mime type; if specified, the body must be of the provided content-type; use this to restrict a route to only use JSON or
  only use URL-encoded input.

### checkKey(value, spec)

Low-level method to validate a specific query or body parameter. Returns `true` or `false`.

### ValidationError

All error reported by this library are of this class. They have the following properties:

- `status`: HTTP status to use (400 or 415)
- `code`: programming-friendly error code, either `E_BAD_PARAM` or `E_BAD_CONTENT_TYPE`
- `key`: the key that caused the error, or `null`
- `message`: error message, suitable for logging but not for displaying to the user (it will not contain sensitive user data, but it might reveal
  information about the app)


HISTORY
0.2.1
=====

* Fixed error message

0.2.0
=====

* Migrated to TypeScript

0.1.0
=====

* First release


test/index.js
// -*- mode: js; indent-tabs-mode: nil; js-basic-offset: 4 -*-
//
// This file is part of Almond Cloud
//
// Copyright 2019 The Board of Trustees of the Leland Stanford Junior University
//
// Author: Giovanni Campagna <gcampagn@cs.stanford.edu>
//
// See COPYING for details

import * as qs from 'qs';
import assert from 'assert';

import * as iv from '../lib';

const TEST_CASES = [
    ['foo=bar', { foo: 'string' }, true],
    ['foo=bar', { foo: 'number' }, false],
    ['foo=bar', { foo: '?string' }, true],
    ['', { foo: '?string' }, true],
    ['', { foo: 'string' }, false],
    ['', { foo: '?number' }, true],

    ['foo=1&bar=2', { foo: 'string', bar: 'number' }, true],
    ['foo=1&bar=lol', { foo: 'string', bar: 'number' }, false],
    ['foo=&bar=2', { foo: 'string', bar: 'number' }, false],

    ['foo=', { foo: '?string' }, true],
    ['foo=', { foo: 'string' }, false],
    ['foo=', { foo: '?number' }, true],
    ['foo=', { foo: 'number' }, false],
    ['foo=', { foo: '?integer' }, true],
    ['foo=', { foo: 'integer' }, false],
    ['foo=', { foo: 'boolean' }, true],
    ['foo=', { foo: '?boolean' }, true],
    ['', { foo: 'boolean' }, true],

    ['foo=1', { foo: 'boolean' }, true],
    ['foo=0', { foo: 'boolean' }, false],
    ['foo=false', { foo: 'boolean' }, false],
    ['foo=true', { foo: 'boolean' }, false],

    ['foo=1', { foo: 'number' }, true],
    ['foo=1.5', { foo: 'number' }, true],
    ['foo=.5e-3', { foo: 'number' }, true],
    ['foo=Infinity', { foo: 'number' }, false],
    ['foo=1', { foo: 'integer' }, true],
    ['foo=1.5', { foo: 'integer' }, false],

    ['foo[]=', { foo: 'array' }, true],
    ['', { foo: 'array' }, false],
    ['foo=1&foo=2', { foo: 'array' }, true],
    ['foo=1', { foo: 'array' }, false],
    ['foo=1', { foo: ['array', 'string'] }, true],
    ['foo=1&foo=2', { foo: ['array', 'string'] }, true],
    ['', { foo: ['array', 'string'] }, false],
    ['', { foo: ['array', '?string'] }, true],
    ['', { foo: '?array' }, true],

    ['foo[key]=1', { foo: 'object' }, true],
    ['foo[key]=1', { foo: 'array' }, false],
    ['foo[key]=1', { foo: 'string' }, false],
    ['foo[]=1&foo[key]=2', { foo: 'array' }, false],

    ['', { foo: 'string' }, false, { json: true }],

    ['class%5B0%5D=foo&class%5B1%5D=bar', { class: '?string' }, false],

    // test that control characters are rejected
    ['foo=\x00', { foo: 'string' }, false],
    ['foo=\x00', { foo: '?string' }, false],
    ['foo=abc\x00def', { foo: 'string' }, false],
    ['foo=abcdef\x00', { foo: 'string' }, false],
    ['foo=\x00abcdef', { foo: 'string' }, false],
    ['foo=\x08', { foo: 'string' }, false],
    ['foo=\x9d', { foo: 'string' }, false],
    ['foo=\x1f', { foo: 'string' }, false],
    ['foo=\x7f', { foo: 'string' }, false],
    ['foo=\x9f', { foo: 'string' }, false],
    // but whitespace is allowed
    ['foo=\n', { foo: 'string' }, true],
    ['foo=\r', { foo: 'string' }, true],
    ['foo=\v', { foo: 'string' }, true],
    ['foo=\f', { foo: 'string' }, true],
    ['foo=\t', { foo: 'string' }, true],
    ['foo=%20', { foo: 'string' }, true],
    ['foo= ', { foo: 'string' }, true],
    ['foo=%0A', { foo: 'string' }, true],

    // Regular Expression:
    ['foo=bar', { foo: /ba[rz]/ }, true],
    ['foo=baz', { foo: /ba[rz]/ }, true],
    ['foo=foo', { foo: /ba[rz]/ }, false],
    ['foo[]=bar&foo[]=baz', { foo: /ba[rz]/ }, false],
    ['foo[]=bar', { foo: /ba[rz]/ }, false],
    ['foo[0]=bar', { foo: /ba[rz]/ }, false],
    ['foo=barbarian', { foo: /ba[rz]/ }, true],
    ['foo=barbarian', { foo: /^ba[rz]$/ }, false],
    ['foo=', { foo: /ba[rz]/ }, false],
    ['foo=', { foo: /^ba[rz]$/ }, false],
    ['foo=', { foo: /^$|^ba[rz]$/ }, true],
    ['', { foo: /ba[rz]/ }, false],
    ['', { foo: /^ba[rz]$/ }, false],
    ['', { foo: /^$|^ba[rz]$/ }, true],
];

function test(i) {
    console.log(`Test Case #${i+1}`);
    const [input, validation, expected, options] = TEST_CASES[i];

    const req = {
        _(x) { return x; },
        query: qs.parse(input)
    };
    const res = {
    };
    const next = (err) => {
        if (err) {
            if (expected)
                assert.fail(`unexpected error for ${input}`);
            assert(err.message.startsWith('missing or invalid parameter'));
            assert.strictEqual(err.code, 'E_BAD_PARAM');
            assert.strictEqual(err.status, 400);
        } else {
            if (!expected)
                assert.fail(`unexpected success for ${input}`);
        }
    };

    iv.validateGET(validation, options)(req, res, next);
}

export default function main() {
    for (let i = 0; i < TEST_CASES.length; i++)
        test(i);
}
if (!module.parent)
    main();


index.js
// -*- mode: js; indent-tabs-mode: nil; js-basic-offset: 4 -*-
//
// This file is part of Thingpedia
//
// Copyright 2019 The Board of Trustees of the Leland Stanford Junior University
//
// Author: Giovanni Campagna <gcampagn@cs.stanford.edu>
//
// See COPYING for details
"use strict";

const typeIs = require('type-is');

class ValidationError extends Error {
    constructor(status, code, key, message) {
        super(message);
        this.status = status;
        this.code = code;
        this.key = key;
    }
}

function checkKey(value, type) {
    if (Array.isArray(type)) {
        for (let option of type) {
            if (checkKey(value, option))
                return true;
        }
        return false;
    }
    if (type instanceof RegExp) {
        if (value === undefined)
            value = '';
        if (typeof value !== 'string')
            return false;
        return type.test(value);
    }

    if (type.startsWith('?')) {
        if (value === undefined || value === '' || value === null)
            return true;
        type = type.substring(1);
    }

    switch (type) {
    case 'null':
        return value === null;

    case 'array':
        return Array.isArray(value);

    case 'string':
        // NOTE: the character ranges U+0000-U+001F and U+007F-U+009F are control
        // characters (NUL, BACKSPACE, DEL, etc.) that have special meaning in many
        // contexts
        // there is no reason to allow them, anywhere
        /* eslint no-control-regex: "off" */
        return typeof value === 'string' && !!value && !/[\x00-\x08\x0e-\x1f\x7f-\x9f]/.test(value);

    case 'boolean':
        // a checkbox is either present (== 1) or absent (== undefined)
        // for api compatibility, also allow true/false
        return value === undefined || value === '1' || value === '' || value === true || value === false;

    // NOTE: parseInt/parseFloat have weird behavior with trailing garbage
    // we don't want to accept that, so we use the unary "+" operator to
    // convert to a number instead
    case 'number':
        return typeof value === 'string' && value !== '' && Number.isFinite(+value);
    case 'integer':
        return typeof value === 'string' && value !== '' && Number.isInteger(+value);

    case 'object':
        return typeof value === 'object' && !Array.isArray(value);

    default:
        // should never be reached
        return false;
    }
}

function failKey(key) {
    return new ValidationError(400 /* Bad Request */, 'E_BAD_PARAM', key, `invalid content-type`);
}

function failContentType() {
    return new ValidationError(415 /* Not Acceptable */, 'E_BAD_CONTENT_TYPE', null, `invalid content-type`);
}

function _validate(req, res, next, body, keys, options) {
    for (let key in keys) {
        if (!checkKey(body[key], keys[key])) {
            next(failKey(key));
            return;
        }
    }
    next();
}

function validateGET(keys, options) {
    options = options || {};
    return function(req, res, next) {
        _validate(req, res, next, req.query, keys, options);
    };
}
function validatePOST(keys, options) {
    options = options || {};

    return function(req, res, next) {
        if (options.accept && !typeIs(req, options.accept)) {
            next(failContentType());
            return;
        }

        _validate(req, res, next, req.body, keys, options);
    };
}

module.exports = {
    ValidationError,

    validateGET,
    validatePOST,

    checkKey,
};


lib/index.ts
// -*- mode: typescript; indent-tabs-mode: nil; js-basic-offset: 4 -*-
//
// This file is part of Thingpedia
//
// Copyright 2019 The Board of Trustees of the Leland Stanford Junior University
//
// Author: Giovanni Campagna <gcampagn@cs.stanford.edu>
//
// See COPYING for details

import typeIs from 'type-is';
import type * as express from 'express';

export class ValidationError extends Error {
    status : number;
    code : string;
    key : string|null;

    constructor(status : number, code : string, key : string|null, message : string) {
        super(message);
        this.status = status;
        this.code = code;
        this.key = key;
    }
}

export type TypeSpec = string[]|string|RegExp;

export function checkKey(value : unknown, type : TypeSpec) : boolean {
    if (Array.isArray(type)) {
        for (const option of type) {
            if (checkKey(value, option))
                return true;
        }
        return false;
    }
    if (type instanceof RegExp) {
        if (value === undefined)
            value = '';
        if (typeof value !== 'string')
            return false;
        return type.test(value);
    }

    if (type.startsWith('?')) {
        if (value === undefined || value === '' || value === null)
            return true;
        type = type.substring(1);
    }

    switch (type) {
    case 'null':
        return value === null;

    case 'array':
        return Array.isArray(value);

    case 'string':
        // NOTE: the character ranges U+0000-U+001F and U+007F-U+009F are control
        // characters (NUL, BACKSPACE, DEL, etc.) that have special meaning in many
        // contexts
        // there is no reason to allow them, anywhere
        /* eslint no-control-regex: "off" */
        return typeof value === 'string' && !!value && !/[\x00-\x08\x0e-\x1f\x7f-\x9f]/.test(value);

    case 'boolean':
        // a checkbox is either present (== 1) or absent (== undefined)
        // for api compatibility, also allow true/false
        return value === undefined || value === '1' || value === '' || value === true || value === false;

    // NOTE: parseInt/parseFloat have weird behavior with trailing garbage
    // we don't want to accept that, so we use the unary "+" operator to
    // convert to a number instead
    case 'number':
        return typeof value === 'string' && value !== '' && Number.isFinite(+value);
    case 'integer':
        return typeof value === 'string' && value !== '' && Number.isInteger(+value);

    case 'object':
        return typeof value === 'object' && !Array.isArray(value);

    default:
        // should never be reached
        return false;
    }
}

function failKey(key : string) : Error {
    return new ValidationError(400 /* Bad Request */, 'E_BAD_PARAM', key, `missing or invalid parameter ${key}`);
}

function failContentType() : Error {
    return new ValidationError(415 /* Not Acceptable */, 'E_BAD_CONTENT_TYPE', null, `invalid content-type`);
}

interface ValidateOptions {
    accept ?: string;
}

function _validate(req : express.Request,
                   res : express.Response,
                   next : (err ?: Error) => void,
                   body : Record<string, unknown>,
                   keys : Record<string, TypeSpec>,
                   options : ValidateOptions) {
    for (const key in keys) {
        if (!checkKey(body[key], keys[key])) {
            next(failKey(key));
            return;
        }
    }
    next();
}

export function validateGET(keys : Record<string, TypeSpec>, options : ValidateOptions = {}) : express.RequestHandler {
    return function(req : express.Request, res : express.Response, next : (err ?: Error) => void) {
        _validate(req, res, next, req.query, keys, options);
    };
}
export function validatePOST(keys : Record<string, TypeSpec>, options : ValidateOptions = {}) : express.RequestHandler {
    return function(req : express.Request, res : express.Response, next : (err ?: Error) => void) {
        if (options.accept && !typeIs(req, options.accept)) {
            next(failContentType());
            return;
        }

        _validate(req, res, next, req.body, keys, options);
    };
}
